order/utils.py

```diff
@@ -14,30 +14,36 @@ def apply_discount(discount, cost):
 def calculate_discounted_cost(obj, *args, **kwargs):
     # the function applies all discounts and selects the one 
     # with the highest benefit until all discounts are over
-    if obj.weight and obj.discount:
+    if obj.weight and obj.discounts:
         excluded_ids = []
-        discounts = obj.discount.all()
+        discounts = obj.discounts.all()
         our_cost = float(obj.cost)
+        loop = True
 
-        while discounts:
+        while discounts and loop:
             main_result = 0
-            main_result_couse_discount = 0
-            course_discount_form = 0
+            main_result_couse_discount = None
+            course_discount_form = None
 
             for discount in discounts:
                 discount.amount = float(discount.amount)
                 result = apply_discount(discount, our_cost)
-                if result > main_result:
+
+                if not result <= 0 and result > main_result:
                     main_result = result
                     main_result_couse_discount = discount.amount
                     course_discount_form = discount.constant_or_percentage
-
-            if course_discount_form == 'constant':
-                our_cost -= main_result_couse_discount
-            elif course_discount_form == 'percentage':
-                our_cost = our_cost - (our_cost * main_result_couse_discount) / 100
-
-            excluded_ids.append(discounts.filter(amount=main_result_couse_discount).values_list('id', flat=True).first())
-            discounts = obj.discount.all().exclude(id__in=excluded_ids)
-
+                    
+                    if course_discount_form == 'constant':
+                        our_cost -= main_result_couse_discount
+                    elif course_discount_form == 'percentage':
+                        our_cost = our_cost - (our_cost * main_result_couse_discount) / 100
+
+                    excluded_ids.append(discounts.filter(amount=main_result_couse_discount).values_list('id', flat=True).first())
+                    discounts = obj.discounts.all().exclude(id__in=excluded_ids)
```

Comment:
You do not need to query all again, you can just exclude from existing queryset as well, discounts.exclude(id=...)